### PR TITLE
Add compression='infer' default to read_csv

### LINKED ISSUE #6929: Add default compression="infer" option to read_csv
As discussed in [cudf#6850](https://github.com/rapidsai/cudf/issues/6850), `dask.dataframe`'s `read_csv` API does not match that of pandas when it comes to compression.  The default setting in pandas is `compression="infer"`, while dask uses `compression=None`.

@dask/io - Is there a strong argument against adding a default `"infer"` option to the dask API?

cc @lmeyerov

### PR DESCRIPTION
**Pull request by rjzamora**
Adds default `compression="infer"` option to `read_csv`.

Closes #6929

**TODO**:
- [x] Test `compression="infer"` for bz2, zip and xz (gzip done)

### CODE REVIEW

**Comment by rjzamora** [comment]
@martindurant - Let me know if you think there is a more elegant way to accomplish what we need here

**Comment by martindurant** [comment]
Maybe `fsspec.utils.infer_compression` does this?

**Comment by martindurant** [comment]
Also, I think if you are using fsspec anyway, you can pass `compression="infer"` to `open_files` and get what you want. I don't remember if any decompression is handled in dask or passed for pandas to handle.

**Comment by rjzamora** [comment]
>Maybe fsspec.utils.infer_compression does this?

Perfect - thanks!

>Also, I think if you are using fsspec anyway, you can pass compression="infer" to open_files and get what you want. I don't remember if any decompression is handled in dask or passed for pandas to handle.

Understood - I'll think about how this relates to the read_csv code path.  We *may* need to detect compression before opening the files so that we are ensuring `blocksize==None`.  Right now, the files are opened within `read_bytes`, and if `blocksize` is not None, this will fail.  I'm not sure it is worthwhile to detect the compression on open if we can use something like `infer_compression`.

**Comment by martindurant** [comment]
https://github.com/databricks/databricks-cli/blob/master/tests/dbfs/test_api.py#L81 ?

**Comment by martindurant** [comment]
(sorry, wrong place for link)

**Comment by rjzamora** [comment]
Thanks for the feedback here @martindurant - We are now using `fsspec.utils.infer_compression`.  I decided not to open any files (with fsspec or pandas) to infer compression, since pandas relies on the file suffix for inference anyway.

Let me know if you have any other thoughts, or feel strongly that we should be doing things differently here.

File: dask/dataframe/io/csv.py [hunk 1]
```diff
@@ -508,6 +511,17 @@ def read_pandas(
     else:
         path_converter = None
 
+    # If compression is "infer", inspect the (first) path suffix and
+    # set the proper compression option if the suffix is recongnized.
+    if compression == "infer":
+        # Translate the input urlpath to a simple path list
+        if not isinstance(urlpath, (str, list, tuple, os.PathLike)):
+            raise TypeError("Path should be a string, os.PathLike, list or tuple")
```
Comment (martindurant) [inline dask/dataframe/io/csv.py:519]: This message isn't quite right.
*IF* we allow other types at all (does file-like work? I thought not), then the message would be "compression='infer' only makes sense when passing a path or set of paths". Or something. I'm not sure we need to have the check at all.

File: dask/dataframe/io/tests/test_csv.py [hunk 2]
```diff
@@ -720,13 +721,15 @@ def test_read_csv_sensitive_to_enforce():
 def test_read_csv_compression(fmt, blocksize):
     if fmt not in compress:
         pytest.skip("compress function not provided for %s" % fmt)
+    suffix = {"gzip": ".gz", "bz2": ".bz2", "zip": ".zip", "xz": ".xz"}.get(fmt, "")
```
Comment (martindurant) [inline dask/dataframe/io/tests/test_csv.py:724]: We should somehow test for no compression.

A comment should be added to say that this test if effectively using "infer" and the extension, or we could test with both infer and explicit (which you already did above, so not necessary).

**Review by martindurant — COMMENTED** [review]
All seems fair, yes. I still have a couple of small comments.

File: dask/dataframe/io/tests/test_csv.py (hunk 2 above)
Comment (rjzamora) [reply dask/dataframe/io/tests/test_csv.py:725]: Okay - I changed the `parametrize` layout a bit and added a non-compression test.  I also added a brief comment to explain that we are relying on `compression="infer"`

File: dask/dataframe/io/csv.py (hunk 1 above)
Comment (rjzamora) [reply dask/dataframe/io/csv.py]: Okay - I think I agree that the instance check is unnecessary.  Removed it

File: dask/dataframe/io/csv.py [hunk 3]
```diff
@@ -508,6 +510,15 @@ def read_pandas(
     else:
         path_converter = None
 
+    # If compression is "infer", inspect the (first) path suffix and
+    # set the proper compression option if the suffix is recongnized.
+    if compression == "infer":
+        # Translate the input urlpath to a simple path list
+        paths = get_fs_token_paths(urlpath, mode="rb", storage_options=kwargs)[2]
```
Comment (ayushdg) [inline dask/dataframe/io/csv.py:517]: Ran into an issue with `read_csv` on a remote source during dask-cudf s3 tests. (https://github.com/rapidsai/cudf/pull/7144)
Looks like storage options isn't being propagated here. 
`paths = get_fs_token_paths(urlpath, mode="rb", storage_options=storage_options)[2]`

Comment (jrbourbeau) [reply dask/dataframe/io/csv.py:517]: Thanks @ayushdg! cc @rjzamora

Comment (rjzamora) [reply dask/dataframe/io/csv.py:517]: Oops! I'll submit a quick fix. Sorry about that!

**Pull request merged**